internal/scheduler/ingester.go

```diff
@@ -0,0 +1,777 @@
+			}
+
+			// If we have more than 1 batch, apply all but the last immedtately.
+			batchesToApply := make([]*Batch, 0)
+			if len(batches) > 1 {
+				batchesToApply = append(batchesToApply, batches[:len(batches)-1]...)
+				batches = []*Batch{batches[len(batches)-1]}
+			}
+
+			// If the most recent batch is large enough,
+			// apply it immedtately to avoid batches becoming too large.
+			if len(batches) == 1 {
+				batch := batches[len(batches)-1]
+				if srv.ShouldWrite(batch) {
+					batchesToApply = append(batchesToApply, batch)
+					batches = make([]*Batch, 0)
+				}
+			}
+
+			// Batches to be applied are forwarded to a separate goroutine responsible for writing to postgres.
+			for _, batch := range batchesToApply {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case batchChannel <- batch:
+				}
+			}
+
+		case <-ticker.C: // Periodically send batches to be applied to postgres.
+			for _, batch := range batches {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case batchChannel <- batch:
+				}
+			}
+			batches = make([]*Batch, 0)
+		}
+	}
+}
+
+func (srv *Ingester) ProcessMessage(ctx context.Context, msg pulsar.Message, batches []*Batch) ([]*Batch, error) {
+	if msg == nil {
+		return batches, nil
+	}
+	log := ctxlogrus.Extract(ctx)
+
+	// Create a fresh batch if we don't already have one.
+	if len(batches) == 0 {
+		batches = append(batches, NewBatch())
+	}
+	batch := batches[len(batches)-1]
+
+	// Store the message id in the batch.
+	// So we can ack messages once they've been written to postgres.
+	//
+	// It's fine to write to postgres and then not ack, since writes are idempotent.
+	// But it's not fine to ack messages not written to postgres,
+	// since skipping messages isn't detected.
+	//
+	// We defer adding the message id to ensure it's added to the last batch created
+	// from this sequence.
+	defer func() {
+		batch := batches[len(batches)-1]
+		batch.MessageIds = append(batch.MessageIds, msg.ID())
+	}()
+
+	// Unmarshal and validate the message.
+	sequence, err := eventutil.UnmarshalEventSequence(ctx, msg.Payload())
+	if err != nil {
+		return batches, err
+	}
+	if sequence == nil || len(sequence.Events) == 0 {
+		return batches, nil
+	}
+
+	// Update the current batch.
+	for i, event := range sequence.GetEvents() {
+		switch e := event.Event.(type) {
+		case *armadaevents.EventSequence_Event_SubmitJob:
+
+			// Filter out jobs not intended for this scheduler.
+			if e.SubmitJob.Scheduler != "pulsar" {
+				continue
+			}
+
+			// If there are job set operations to be applied,
+			// we can't add new jobs for any affected job sets to this batch.
+			// Since jobs submitted after those operations should not be affected.
+			// To that end, we create new batches as necessary here.
+			if _, ok := batch.JobSetsCancelled[sequence.GetJobSetName()]; ok {
+				batches = append(batches, NewBatch())
```

Comment:
this is a minor thing but isn't it clearer to do:
```
batch = NewBatch()
batches = append(batches, batch)
```
This has the advantage of a) fewer operations and b) each line does only one thing.